File: app/src/main/java/com/github/overpass/gather/model/repo/upload/UploadImageRepo.java
package com.github.overpass.gather.model.repo.upload;

import android.content.ContentResolver;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.provider.MediaStore;

import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.github.overpass.gather.model.data.FailedTask;
import com.github.overpass.gather.screen.auth.register.add.ImageUploadStatus;
import com.github.overpass.gather.model.commons.Runners;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.io.ByteArrayOutputStream;

public class UploadImageRepo {

    public static final String BUCKET_AVATARS = "Avatars";
    public static final String BUCKET_MEETINGS = "Meetings";

    private final FirebaseStorage storage;

    public UploadImageRepo(FirebaseStorage storage) {
        this.storage = storage;
    }

    public LiveData<ImageUploadStatus> saveImage(ContentResolver contentResolver,
                                                 Uri imageUri,
                                                 String bucket,
                                                 String userId,
                                                 String imageName) {
        MutableLiveData<ImageUploadStatus> data = new MutableLiveData<>();
        data.setValue(new ImageUploadStatus.Progress(0));
        Runners.io().execute(() -> {
            String path = bucket + "/" + userId + "/" + imageName + ".jpg";
            StorageReference storageReference = storage.getReference().child(path);
            storageReference.putBytes(uriToBytes(contentResolver, imageUri))
                    .addOnFailureListener(e -> data.postValue(new ImageUploadStatus.Error(e)))
                    .addOnProgressListener(temp -> {
                        long percent = temp.getBytesTransferred() / temp.getTotalByteCount();
                        data.postValue(new ImageUploadStatus.Progress(percent));
                    })
                    .onSuccessTask(Runners.io(), docRef -> storageReference.getDownloadUrl())
                    .addOnSuccessListener(uri -> {
                        data.postValue(new ImageUploadStatus.Success(uri));
                    })
                    .addOnFailureListener(e -> {
                        data.postValue(new ImageUploadStatus.Error(e));
                    });
        });
        return data;
    }

    @WorkerThread
    public byte[] uriToBytes(ContentResolver contentResolver, Uri imageUri) {
        Bitmap bitmap = uriToBitmap(contentResolver, imageUri);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);
        return baos.toByteArray();
    }

    @WorkerThread
    public Bitmap uriToBitmap(ContentResolver contentResolver, Uri imageUri) {
        String[] filePathColumn = {MediaStore.Images.Media.DATA};
        // Get the cursor
        Cursor cursor = contentResolver.query(imageUri, filePathColumn, null,
                null, null);
        // Move to first row
        cursor.moveToFirst();
        //Get the column index of MediaStore.Images.Media.DATA
        int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
        //Gets the String value in the column
        String imgDecodableString = cursor.getString(columnIndex);
        cursor.close();
        // Set the Image in ImageView after decoding the String
        return BitmapFactory.decodeFile(imgDecodableString);
    }
}

File: app/src/main/java/com/github/overpass/gather/model/repo/upload/UploadImageRepo.kt
package com.github.overpass.gather.model.repo.upload

import android.content.ContentResolver
import android.net.Uri
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.github.overpass.gather.model.commons.Result
import com.github.overpass.gather.model.commons.Runners
import com.github.overpass.gather.model.commons.image.ImageConverter
import com.github.overpass.gather.screen.auth.register.add.ImageUploadStatus
import com.google.android.gms.tasks.SuccessContinuation
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow

class UploadImageRepo(
        private val storage: FirebaseStorage,
        private val imageConverter: ImageConverter
) {

    @Deprecated("""Use this:
        fun saveImage(imageBytes: ByteArray,
                  bucket: String,
                  folder: String,
                  imageName: String
        ): Task<ImageUploadStatus>
    """)
    fun saveImage(contentResolver: ContentResolver,
                  imageUri: Uri,
                  bucket: String,
                  folder: String,
                  imageName: String): LiveData<ImageUploadStatus> {
        val data = MutableLiveData<ImageUploadStatus>()
        data.setValue(ImageUploadStatus.Progress(0))
        Runners.io().execute {
            val path = "$bucket/$folder/$imageName$IMAGE_EXTENSION"
            val storageReference = storage.reference.child(path)
            val imageBytes = imageConverter.resolveBytesFromImageUri(imageUri)
            storageReference.putBytes(imageBytes)
                    .addOnFailureListener { e -> data.postValue(ImageUploadStatus.Error(e)) }
                    .addOnProgressListener { temp ->
                        val percent = temp.bytesTransferred / temp.totalByteCount
                        data.postValue(ImageUploadStatus.Progress(percent))
                    }
                    .onSuccessTask<Uri>(Runners.io(), SuccessContinuation { docRef ->
                        storageReference.downloadUrl
                    })
                    .addOnSuccessListener { uri -> data.postValue(ImageUploadStatus.Success(uri)) }
                    .addOnFailureListener { e -> data.postValue(ImageUploadStatus.Error(e)) }
        }
        return data
    }

    @ExperimentalCoroutinesApi
    fun saveImage2(
            imageBytes: ByteArray,
            bucket: String,
            folder: String,
            imageName: String
    ): Flow<Result<String>> = callbackFlow {
        send(Result.Loading())
        val path = "$bucket/$folder/$imageName$IMAGE_EXTENSION"
        val storageReference = storage.reference.child(path)
        storageReference.putBytes(imageBytes)
                .addOnProgressListener { temp ->
                    val current = temp.bytesTransferred.toDouble()
                    val total = temp.totalByteCount.toDouble()
                    val percent = (current / total * 100).toInt()
                    offer(Result.Loading(percent))
                }
                .onSuccessTask { storageReference.downloadUrl }
                .addOnSuccessListener { offer(Result.Success(it.toString())) }
                .addOnFailureListener { offer(Result.Error(it)) }
        awaitClose()
    }

    companion object {
        const val IMAGE_EXTENSION = ".jpg"
        const val BUCKET_AVATARS = "Avatars"
        const val BUCKET_MEETINGS = "Meetings"
    }
}
